File: server-go/internal/service/email_service.go
package service

import (
	"crypto/rand"
	"fmt"
	"math/big"
	"time"

	"github.com/yourusername/gratitude-journal-api/internal/config"
	"github.com/yourusername/gratitude-journal-api/internal/models"
	"github.com/yourusername/gratitude-journal-api/internal/repository"
	"gopkg.in/gomail.v2"
)

type EmailService struct {
	config   *config.EmailConfig
	repo     *repository.EmailVerificationRepository
	dialer   *gomail.Dialer
}

func NewEmailService(cfg *config.EmailConfig) *EmailService {
	dialer := gomail.NewDialer(cfg.Host, cfg.Port, cfg.User, cfg.Password)
	return &EmailService{
		config: cfg,
		repo:   repository.NewEmailVerificationRepository(),
		dialer: dialer,
	}
}

// GenerateVerificationCode generates a 6-digit verification code
func (s *EmailService) GenerateVerificationCode() (string, error) {
	max := big.NewInt(1000000)
	n, err := rand.Int(rand.Reader, max)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%06d", n.Int64()), nil
}

// SendVerificationCode sends a verification code to the email
func (s *EmailService) SendVerificationCode(email, language string) error {
	// Generate code
	code, err := s.GenerateVerificationCode()
	if err != nil {
		return fmt.Errorf("failed to generate code: %w", err)
	}

	// Delete any existing codes for this email
	_ = s.repo.DeleteVerificationsByEmail(email)

	// Save verification to database
	verification := &models.EmailVerification{
		Email:     email,
		Code:      code,
		ExpiresAt: time.Now().Add(5 * time.Minute),
		Used:      false,
	}
	if err := s.repo.CreateVerification(verification); err != nil {
		return fmt.Errorf("failed to save verification: %w", err)
	}

	// Send email
	if err := s.sendEmail(email, code, language); err != nil {
		return fmt.Errorf("failed to send email: %w", err)
	}

	return nil
}

// VerifyCode verifies the code and returns true if valid
func (s *EmailService) VerifyCode(email, code string) (bool, error) {
	verification, err := s.repo.GetValidVerification(email, code)
	if err != nil {
		return false, nil // Invalid code
	}

	// Mark as used
	if err := s.repo.MarkAsUsed(verification.ID); err != nil {
		return false, fmt.Errorf("failed to mark verification as used: %w", err)
	}

	return true, nil
}

func (s *EmailService) sendEmail(to, code, language string) error {
	m := gomail.NewMessage()
	m.SetHeader("From", s.config.From)
	m.SetHeader("To", to)

	var subject, body string
	if language == "en" {
		subject = "Awaken - Your Verification Code"
		body = s.getEnglishEmailBody(code)
	} else {
		subject = "觉醒日志 - 您的验证码"
		body = s.getChineseEmailBody(code)
	}

	m.SetHeader("Subject", subject)
	m.SetBody("text/html", body)

	return s.dialer.DialAndSend(m)
}

func (s *EmailService) getChineseEmailBody(code string) string {
	return fmt.Sprintf(`
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; background: #f5f5f5; padding: 20px; }
        .container { max-width: 500px; margin: 0 auto; background: white; border-radius: 16px; padding: 40px; box-shadow: 0 4px 20px rgba(0,0,0,0.1); }
        .logo { text-align: center; margin-bottom: 30px; }
        .logo-text { font-size: 28px; font-weight: bold; color: #FF8C00; }
        .title { font-size: 20px; color: #333; margin-bottom: 20px; text-align: center; }
        .code-box { background: linear-gradient(135deg, #FF8C00, #FFA500); border-radius: 12px; padding: 20px; text-align: center; margin: 30px 0; }
        .code { font-size: 36px; font-weight: bold; color: white; letter-spacing: 8px; }
        .note { color: #666; font-size: 14px; text-align: center; margin-top: 20px; }
        .footer { color: #999; font-size: 12px; text-align: center; margin-top: 30px; padding-top: 20px; border-top: 1px solid #eee; }
    </style>
</head>
<body>
    <div class="container">
        <div class="logo">
            <span class="logo-text">🌅 觉醒日志</span>
        </div>
        <div class="title">您的登录验证码</div>
        <div class="code-box">
            <div class="code">%s</div>
        </div>
        <div class="note">
            验证码有效期为 5 分钟<br>
            如果这不是您的操作，请忽略此邮件
        </div>
        <div class="footer">
            © 2025 Awaken Journal · 觉醒日志
        </div>
    </div>
</body>
</html>
`, code)
}

func (s *EmailService) getEnglishEmailBody(code string) string {
	return fmt.Sprintf(`
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; background: #f5f5f5; padding: 20px; }
        .container { max-width: 500px; margin: 0 auto; background: white; border-radius: 16px; padding: 40px; box-shadow: 0 4px 20px rgba(0,0,0,0.1); }
        .logo { text-align: center; margin-bottom: 30px; }
        .logo-text { font-size: 28px; font-weight: bold; color: #FF8C00; }
        .title { font-size: 20px; color: #333; margin-bottom: 20px; text-align: center; }
        .code-box { background: linear-gradient(135deg, #FF8C00, #FFA500); border-radius: 12px; padding: 20px; text-align: center; margin: 30px 0; }
        .code { font-size: 36px; font-weight: bold; color: white; letter-spacing: 8px; }
        .note { color: #666; font-size: 14px; text-align: center; margin-top: 20px; }
        .footer { color: #999; font-size: 12px; text-align: center; margin-top: 30px; padding-top: 20px; border-top: 1px solid #eee; }
    </style>
</head>
<body>
    <div class="container">
        <div class="logo">
            <span class="logo-text">🌅 Awaken</span>
        </div>
        <div class="title">Your Verification Code</div>
        <div class="code-box">
            <div class="code">%s</div>
        </div>
        <div class="note">
            This code expires in 5 minutes<br>
            If you didn't request this, please ignore this email
        </div>
        <div class="footer">
            © 2025 Awaken Journal
        </div>
    </div>
</body>
</html>
`, code)
}

// CleanupExpiredVerifications removes expired verification codes
func (s *EmailService) CleanupExpiredVerifications() error {
	return s.repo.DeleteExpiredVerifications()
}

File: packages/server/_core/email-service.ts
import nodemailer from 'nodemailer';

// Email configuration from environment variables
const EMAIL_HOST = process.env.EMAIL_HOST || 'smtp.gmail.com';
const EMAIL_PORT = parseInt(process.env.EMAIL_PORT || '587');
const EMAIL_USER = process.env.EMAIL_USER;
const EMAIL_PASS = process.env.EMAIL_PASS;
const EMAIL_FROM = process.env.EMAIL_FROM || EMAIL_USER;

// Create reusable transporter
let transporter: nodemailer.Transporter | null = null;

function getTransporter() {
  if (!EMAIL_USER || !EMAIL_PASS) {
    console.warn('[Email] Email credentials not configured. Emails will be logged to console only.');
    return null;
  }

  if (!transporter) {
    transporter = nodemailer.createTransport({
      host: EMAIL_HOST,
      port: EMAIL_PORT,
      secure: EMAIL_PORT === 465, // true for 465, false for other ports
      auth: {
        user: EMAIL_USER,
        pass: EMAIL_PASS,
      },
    });
  }

  return transporter;
}

/**
 * Send verification code email
 */
export async function sendVerificationCodeEmail(
  to: string,
  code: string,
  language: 'zh' | 'en' = 'zh'
): Promise<boolean> {
  const subject = language === 'zh' ? '您的验证码' : 'Your Verification Code';
  
  const htmlContent = language === 'zh' ? `
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { text-align: center; padding: 20px 0; border-bottom: 2px solid #0a7ea4; }
        .content { padding: 30px 0; }
        .code-box { background: #f5f5f5; border: 2px dashed #0a7ea4; border-radius: 8px; padding: 20px; text-align: center; margin: 20px 0; }
        .code { font-size: 32px; font-weight: bold; color: #0a7ea4; letter-spacing: 8px; }
        .footer { text-align: center; padding: 20px 0; border-top: 1px solid #eee; color: #999; font-size: 12px; }
      </style>
    </head>
    <body>
      <div class="container">
        <div class="header">
          <h1 style="color: #0a7ea4; margin: 0;">感恩日记</h1>
        </div>
        <div class="content">
          <p>您好，</p>
          <p>您正在登录感恩日记应用，您的验证码是：</p>
          <div class="code-box">
            <div class="code">${code}</div>
          </div>
          <p>验证码有效期为 <strong>10分钟</strong>，请尽快使用。</p>
          <p>如果这不是您的操作，请忽略此邮件。</p>
        </div>
        <div class="footer">
          <p>此邮件由系统自动发送，请勿回复。</p>
          <p>© 2026 感恩日记 Gratitude Journal</p>
        </div>
      </div>
    </body>
    </html>
  ` : `
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Arial, sans-serif; line-height: 1.6; color: #333; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { text-align: center; padding: 20px 0; border-bottom: 2px solid #0a7ea4; }
        .content { padding: 30px 0; }
        .code-box { background: #f5f5f5; border: 2px dashed #0a7ea4; border-radius: 8px; padding: 20px; text-align: center; margin: 20px 0; }
        .code { font-size: 32px; font-weight: bold; color: #0a7ea4; letter-spacing: 8px; }
        .footer { text-align: center; padding: 20px 0; border-top: 1px solid #eee; color: #999; font-size: 12px; }
      </style>
    </head>
    <body>
      <div class="container">
        <div class="header">
          <h1 style="color: #0a7ea4; margin: 0;">Gratitude Journal</h1>
        </div>
        <div class="content">
          <p>Hello,</p>
          <p>You are logging into Gratitude Journal. Your verification code is:</p>
          <div class="code-box">
            <div class="code">${code}</div>
          </div>
          <p>This code will expire in <strong>10 minutes</strong>. Please use it soon.</p>
          <p>If you didn't request this, please ignore this email.</p>
        </div>
        <div class="footer">
          <p>This is an automated email. Please do not reply.</p>
          <p>© 2026 Gratitude Journal</p>
        </div>
      </div>
    </body>
    </html>
  `;

  const textContent = language === 'zh' 
    ? `您的验证码是：${code}\n\n验证码有效期为10分钟，请尽快使用。\n\n如果这不是您的操作，请忽略此邮件。`
    : `Your verification code is: ${code}\n\nThis code will expire in 10 minutes. Please use it soon.\n\nIf you didn't request this, please ignore this email.`;

  const transport = getTransporter();

  // If no transporter configured, log to console
  if (!transport) {
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log('📧 EMAIL (Console Mode - No SMTP configured)');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`To: ${to}`);
    console.log(`Subject: ${subject}`);
    console.log(`Verification Code: ${code}`);
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    return true;
  }

  // Send actual email
  try {
    const info = await transport.sendMail({
      from: `"感恩日记 Gratitude Journal" <${EMAIL_FROM}>`,
      to,
      subject,
      text: textContent,
      html: htmlContent,
    });

    console.log(`[Email] Verification code sent to ${to}: ${info.messageId}`);
    return true;
  } catch (error) {
    console.error('[Email] Failed to send verification code:', error);
    // Fallback to console logging
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log('📧 EMAIL (Fallback - SMTP failed)');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`To: ${to}`);
    console.log(`Subject: ${subject}`);
    console.log(`Verification Code: ${code}`);
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    return true;
  }
}
